### PR TITLE
6 add filtering for school level

### PR DESCRIPTION
**Pull request by katiebuntic**
hey both :)

I have added in:
- filtering through so only RR Teachers are shown
- filtering through the tables as explained [here](https://github.com/UCL-ARC/rred-reports/issues/6)
- added some test pupils to the masterfile with various date cases to check the filters work (school_id = RRS2030220)

still need to:
- write a proper test, tried to have a go at `assert` now but wasn't quite getting it so will have another go tomorrow
- remove test pupils (wasn't quite sure what this meant though as the whole file is a test - is this something related to the redcap data?)

### CODE REVIEW

File: src/rred_reports/reports/schools.py [hunk 1]
```diff
@@ -70,51 +70,54 @@
...
+    ]
+    return teacher_filtered[teacher_filtered.school_id == school_id].copy()
 
-def filter_three(school_dataframe: pd.DataFrame):
-    """Filter for table three, for later use
 
-    Args: school_filter(pd.DataFrame)"""
-    return school_dataframe
+# specific filter for each table
+def filter_for_one_two_five(school_dataframe: pd.DataFrame):
```
Comment (stefpiatek) [inline src/rred_reports/reports/schools.py:87]: all of the filter functions should take a year and then we do the dates based on that year! The top level function will also need to have a year passed in too

File: src/rred_reports/reports/schools.py [hunk 2]
```diff
@@ -70,51 +70,54 @@
...
 
-def filter_five(school_dataframe: pd.DataFrame):
-    """Filter for table five, for later use
+    Args: school_filter(pd.DataFrame)
 
-    Args: school_filter(pd.DataFrame)"""
-    return school_dataframe
+    Returns: school_filter(pd.DataFrame) filtered by exit_outcome and exit_date
+    """
+    return school_dataframe[(school_dataframe["exit_outcome"] == "Discontinued") | (school_dataframe["exit_outcome"] == "Referred to school")][
```
Comment (stefpiatek) [inline src/rred_reports/reports/schools.py:108]: I think we could use `.isin` here to make it a bit more clear, and same for other other times. 
```suggestion
    return school_dataframe[(school_dataframe["exit_outcome"].isin(["Discontinued", "Referred to school"]) & 

```

File: tests/test_report_writing.py [hunk 3]
```diff
@@ -38,3 +40,27 @@ def test_school_name_replaced_in_paragraphs(data_path, templates_dir, temp_out_d
...
+    Table 1, 2, 5: test_1_2021-22, test_3_2021-22, test_4_2021-22, test_5_2021-22, test_6_2021-22
+    Table 3&4: test_1_2021-22, test_4_2021-22
+    Table 6: No data
+    """
+
+    nested_data = parse_masterfile(data_path / "example_masterfile.xlsx")
+    testing_df = join_masterfile_dfs(nested_data)
+    test_pupils_school_data = school_filter(testing_df, "RRS2030220")
+
+    # assert test_pupils_school_data [(test_pupils_school_data.reg_rr_title != "Teacher Leader") |
```
Comment (stefpiatek) [inline tests/test_report_writing.py:59]: Might be worth having a look at our other tests for how we assert specific values, but also  tap @harryjmoss if you're blocked on it

**Review by stefpiatek — COMMENTED** [review]
Looking good, we want to make the time filtering not fixed to one year so we can use these more generically and some other suggestions

**Review by stefpiatek — COMMENTED** [review]
Ah yes you're right, filtering out the test participants should happen in the wide to long extract stage, so you can ignore that. Thanks!

Some suggestions for you!

File: src/rred_reports/reports/schools.py [hunk 4]
```diff
@@ -69,52 +70,76 @@
...
+    """
+    report_start, report_end = set_dates(report_year)
+
+    outcome_filtered = school_dataframe[(school_dataframe["exit_outcome"].isin(["Discontinued", "Referred to school"]))]
 
+    return outcome_filtered[(outcome_filtered["exit_date"] > report_start) & (outcome_filtered["exit_date"] < report_end)]
 
-def filter_six(school_dataframe: pd.DataFrame):
+
+def filter_six(school_dataframe: pd.DataFrame, report_year):
```
Comment (harryjmoss) [inline src/rred_reports/reports/schools.py:130]: Does the docstring need updating here?

File: src/rred_reports/reports/schools.py [hunk 5]
```diff
@@ -195,21 +220,23 @@ def populate_school_tables(school_df: pd.DataFrame, template_path: Path) -> Temp
...
 
 
-def populate_school_data(school_df: pd.DataFrame, template_path: Path, school_id, output_path=None, school_placeholder="School A") -> TemplateFiller:
+def populate_school_data(
+    school_df: pd.DataFrame, template_path: Path, school_id, report_year, output_path=None, school_placeholder="School A"
+) -> TemplateFiller:
     """Function to populate and save the template with: name of school and filled tables
 
-    Args: school_filter(pd.DataFrame), Path, school_id
+    Args: school_filter(pd.DataFrame), Path, school_id, report_year
```
Comment (harryjmoss) [inline src/rred_reports/reports/schools.py:235]: Looks like we're missing `output_path` and `school_placeholder` here

File: src/rred_reports/reports/schools.py
Comment (harryjmoss) [inline src/rred_reports/reports/schools.py:1]: Looks good! Couple of very minor things with the docstrings and would be good to include the return type in the function definitions

Comment (katiebuntic) [reply src/rred_reports/reports/schools.py:1]: ah yes! thanks for the comments @harryjmoss update them now :)

File: tests/data/~$example_masterfile.xlsx
Comment (harryjmoss) [inline tests/data/~$example_masterfile.xlsx:1]: I think we can remove this temp file

File: tests/test_report_writing.py [hunk 6]
```diff
@@ -30,11 +47,89 @@ def test_school_name_replaced_in_paragraphs(data_path, templates_dir, temp_out_d
...
+
+    # test_5_2021-22, Incomplete and within date boundary. Should be in 1,2,5.
+    pupil_5 = test_filter_for_one_two_five.loc[test_filter_for_one_two_five.pupil_no == "test_5_2021-22"]
+    assert ((pupil_5["entry_date"] > "2022-8-1") | (pupil_5["entry_date"] < "2021-7-31")).all
+    assert ((pupil_5["exit_date"] < "2022-8-1") | (pupil_5["exit_date"] > "2021-7-31") | (pupil_5["exit_date"].isna())).all()
+
+    # test_6_2021-22, Ongoing and entry date within boundary and NA exit date. Should be in 1,2,5.
+    pupil_6 = test_filter_for_one_two_five.loc[test_filter_for_one_two_five.pupil_no == "test_6_2021-22"]
+    assert ((pupil_6["entry_date"] < "2022-8-1") | (pupil_6["entry_date"] > "2021-7-31")).all()
+    assert (pupil_6["exit_date"].isna()).all()
```
Comment (harryjmoss) [inline tests/test_report_writing.py:103]: might make sense to make these separate tests (maybe with `test_filter_for_one_two_five` as a fixture?)

**Review by stefpiatek — APPROVED** [review]
Thanks for getting this in, lovely to see it all coming together

**Review by harryjmoss — APPROVED** [review]
Some v minor comments but looks good!

File: tests/test_report_writing.py (hunk 6 above)
Comment (stefpiatek) [inline tests/test_report_writing.py:103]: Ooh I think maybe parameterise the positive an negative tests, I think something like this should be work
```suggestion
@pytest.mark.parametrize("pupil_id",
                         # test_1_2021-22, Discontinued and within date boundary, no month3 or month 6 results. Should show in table 1,2,3,4,5.
                         ["test_1_2021-22",
                          # test_3_2021-22, Discontinued, entry date IN report date but exit date OUT of report date, no month3 or month 6 results. Should be in 1,2,5.
                          "test_3_2021-22",
                          # test_4_2021-22, Discontinued, entry date OUT of report date but exit date IN report date, month3 testdate in range. Should be in all tables.
                          "test_4_2021-22",
                          # test_5_2021-22, Incomplete and within date boundary. Should be in 1,2,5.
                          "test_5_2021-22",
                          # test_6_2021-22, Ongoing and entry date within boundary and NA exit date. Should be in 1,2,5.
                          "test_6_2021-22"])
def test_first_filter_kept(example_school_data, pupil_id):
    """
    Given a school that has data that is not within the date range
    When the template is populated with redcap data
    Then this template should be filled with the following:
    Table 1, 2, 5: test_1_2021-22, test_3_2021-22, test_4_2021-22, test_5_2021-22, test_6_2021-22
    """
    test_filter_for_one_two_five = filter_for_one_two_five(example_school_data, 2022)
    # first filter test
    assert (~example_school_data["reg_rr_title"].isin(["Teacher Leader", "Teacher Leader Only", "Teacher Leader + Support Role"])).any()
    assert (example_school_data["entry_date"] < "2022-8-1").all()
    assert (
        (example_school_data["exit_date"].isna()) | (example_school_data["exit_date"] > "2021-7-31") | (example_school_data["exit_date"] < "2022-8-1")
    ).all()

    pupil = test_filter_for_one_two_five.loc[test_filter_for_one_two_five.pupil_no == "test_1_2021-22"]
    assert ((pupil["entry_date"] < "2022-8-1") | (pupil["entry_date"] > "2021-7-31")).all()
    assert ((pupil["exit_date"] < "2022-8-1") | (pupil["exit_date"] > "2021-7-31") | (pupil["exit_date"].isna())).all()

def test_first_filter_removed(example_school_data):
    """
    Given a school that has data that is not within the date range
    When the template is populated with redcap data
    Then this template should not have pupil test_2_2021-22
    """
    test_filter_for_one_two_five = filter_for_one_two_five(example_school_data, 2022)

    # test_2_2021-22, Discontinued, not within date boundary, month3 testdate out of range, no month6. Should not be in any tables.
    pupil_2 = test_filter_for_one_two_five.loc[test_filter_for_one_two_five.pupil_no == "test_2_2021-22"]
    assert ((~pupil_2["entry_date"] < "2022-8-1") | (~pupil_2["entry_date"] > "2021-7-31")).all()
    assert ((~pupil_2["exit_date"] < "2022-8-1") | (~pupil_2["exit_date"] > "2021-7-31") | (pupil_2["exit_date"].isna())).all()

```

Comment (katiebuntic) [reply tests/test_report_writing.py:103]: ah I merged without commiting this in! will commit now

**Pull request merged**